src/Sparrow/Allocators.PointerTypes.cs

```diff
@@ -0,0 +1,491 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Sparrow
+{
+    public interface IPointerType { }
+    public interface IPointerType<T> : IPointerType where T : struct { }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public readonly unsafe struct Pointer : IPointerType, Meta.IDescribe
+    {
+        public readonly void* Ptr;
+
+        public readonly int Size;
+
+        public Pointer(void* ptr, int size)
+        {
+            this.Ptr = ptr;
+            this.Size = size;
+        }
+
+        public bool IsValid
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get { return Ptr != null; }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<byte> AsSpan()
+        {
+            return new Span<byte>(Ptr, (int)Size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<byte> AsSpan(int length)
+        {
+#if VALIDATE
+            if (length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+#endif
+
+            return new Span<byte>(Ptr, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<byte> AsSpan(int start, int length)
+        {
+#if VALIDATE
+            if (start + length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+#endif
+
+            return new Span<byte>((byte*)Ptr + start, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<byte> AsReadOnlySpan()
+        {
+            return new ReadOnlySpan<byte>(Ptr, (int)Size);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<byte> AsReadOnlySpan(int length)
+        {
+#if VALIDATE
+            if (length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
+#endif
+
+            return new ReadOnlySpan<byte>(Ptr, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ReadOnlySpan<byte> AsReadOnlySpan(int start, int length)
+        {
+#if VALIDATE
+            if (start + length > Size)
+                throw new ArgumentException($"{nameof(length)} cannot be bigger than block size.");            
```

Comment:
We do, we shouldn't (potentially) kill the process because of a leak... we should degrade 'gracefully' instead. That is the reason why I set it to die horribly in debug instead.

Reply:
Makes sense, better to check than touch memory that is potentially not ours... I originally though it was the Allocator Leaks detection. In this case, you are right.